go/adbc/driver/bigquery/connection.go

```diff
@@ -0,0 +1,868 @@
+		if structType == nil {
+			return arrow.Field{}, adbc.Error{
+				Code: adbc.StatusInvalidArgument,
+				Msg:  fmt.Sprintf("Cannot create a struct schema for record `%s`", fieldRecords),
+			}
+		}
+		field.Type = structType
+	default:
+		return arrow.Field{}, adbc.Error{
+			Code: adbc.StatusInvalidArgument,
+			Msg:  fmt.Sprintf("Cannot handle data type `%s`, type=`%s`", dataType, typeString),
+		}
+	}
+	return field, nil
+}
+
+func parsePrecisionAndScale(name, typeString string) (int32, int32, error) {
+	typeString = strings.TrimSpace(typeString)
+	if len(typeString) == 0 {
+		return 0, 0, adbc.Error{
+			Code: adbc.StatusInvalidData,
+			Msg:  fmt.Sprintf("Cannot parse data type `%s` for field `%s`", typeString, name),
+		}
+	}
+
+	if typeString == "NUMERIC" {
+		return 38, 9, nil
+	} else if typeString == "BIGNUMERIC" {
+		return 76, 38, nil
+	}
+
+	values := strings.Split(strings.TrimRight(typeString[strings.Index(typeString, "(")+1:], ")"), ",")
+	if len(values) != 2 {
+		return 0, 0, adbc.Error{
+			Code: adbc.StatusInvalidData,
+			Msg:  fmt.Sprintf("Cannot parse data type `%s` for field `%s`", typeString, name),
+		}
+	}
+
+	precision, err := strconv.ParseInt(values[0], 10, 32)
+	if err != nil {
+		return 0, 0, adbc.Error{
+			Code: adbc.StatusInvalidData,
+			Msg:  fmt.Sprintf("Cannot parse precision `%s`: %s", values[0], err.Error()),
+		}
+	}
+
+	scale, err := strconv.ParseInt(values[1], 10, 32)
+	if err != nil {
+		return 0, 0, adbc.Error{
+			Code: adbc.StatusInvalidData,
+			Msg:  fmt.Sprintf("Cannot parse scale `%s`: %s", values[1], err.Error()),
+		}
+	}
+	return int32(precision), int32(scale), nil
+}
+
+func patternToRegexp(pattern *string) (*regexp.Regexp, error) {
+	patternString := ""
+	if pattern != nil {
+		patternString = *pattern
+	}
+	patternString = strings.TrimSpace(patternString)
+
+	convertedPattern := ".*"
+	if patternString != "" {
+		convertedPattern = fmt.Sprintf("(?i)^%s$", strings.ReplaceAll(strings.ReplaceAll(patternString, "_", "."), "%", ".*"))
+	}
+	r, err := regexp.Compile(convertedPattern)
+	if err != nil {
+		return nil, adbc.Error{
+			Code: adbc.StatusInvalidArgument,
+			Msg:  fmt.Sprintf("Cannot parse pattern `%s`: %s", patternString, err.Error()),
+		}
+	}
+	return r, nil
+}
+
+func (c *connectionImpl) getDatasetsInProject(ctx context.Context, projectIDPattern, datasetsIDPattern *string, cb FilteredDatasetHandler) error {
+	pattern, err := patternToRegexp(projectIDPattern)
+	if err != nil {
+		return err
+	}
+	if !pattern.MatchString(c.client.Project()) {
+		return iterator.Done
+	}
+
+	pattern, err = patternToRegexp(datasetsIDPattern)
+	if err != nil {
+		return err
+	}
+
+	it := c.client.Datasets(ctx)
+	for {
+		dataset, err := it.Next()
+		if err != nil {
+			if errors.Is(err, iterator.Done) {
+				break
+			}
+			return err
+		}
+
+		if pattern.MatchString(dataset.DatasetID) {
+			err = cb(dataset)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
```

Comment:
is this really the only way we can do this? we can't just make a sql query or otherwise get this information from bigquery, we have to iterate and perform the pattern matching ourselves?

Reply:
It's possible if the user moves all datasets to a single location, otherwise we have to do multiple SQL queries to actually get all datasets. Currently there're 13 regions in the Americas, 11 regions in Asia Pacific, 12 regions in Europe, 3 regions in the Middle East and 1 region in Africa, a total 40 regions in the world ([reference](https://cloud.google.com/bigquery/docs/locations)). And that would be 40 SQL queries to get all datasets in one project because sadly that BigQuery doesn't support cross-region queries even if we're only interested in just some metadata, which means that we cannot do something like

```sql
SELECT
  * 
FROM 
  region-us-central1.INFORMATION_SCHEMA.SCHEMATA
UNION ALL 
SELECT 
  *
FROM
  region-us-west1.INFORMATION_SCHEMA.SCHEMATA;
```

because it would result in errors

<img width="1130" alt="Screenshot 2024-04-25 at 06 29 20" src="https://github.com/apache/arrow-adbc/assets/89497197/eed72cdb-75c9-43a6-8d43-70d88da36e71">

Also, their [example](https://cloud.google.com/bigquery/docs/information-schema-datasets-schemata#scope_and_syntax) says you can do 

```sql
SELECT * FROM region-us.INFORMATION_SCHEMA.SCHEMATA;
```

And it appears that it can magically retrieve all datasets in the US region but It's not the case. Because that would only count the datasets in the [multi regional locations](https://cloud.google.com/bigquery/docs/locations#multi-regional-locations).